### ISSUE TITLE
[ottf] add stack trace logging feature to aid in manufacturing test failures

### ISSUE THREAD

**Issue by timothytrippel** [lowRISC/opentitan#17703]
With #17495, we are refactoring the testutils to return `status_t`s to enable gracefully unwinding the stack when test errors are encountered in a test. This allows us to print a test result in JSON format, in a loop over the OTTF console so the ATE host can pick up the error. However, if the same function is called in many places throughout a test, it may be hard to determine the exact failure point.

Maintaining a lightweight, fixed-size, stack trace to print to the console on a test failure would be simple way to address this issue.

- [x] investigate how this feature can be added to the OTTF console library
- [x] implement mechanism in OTTF console library

**Comment by timothytrippel** [comment]
This ^ was partly how I thought about approaching this. Another piece of the puzzle (I think) would be to maintain a global, fixed-sized, queue/stack in the OTTF somewhere that the code embedded in the TRY macro could push call-sites too. I say queue/stack, because in practice it will be easier to just maintain a call stack of X-deep calls, and we most likely want a FIFO behavior when filling it, but  FILO when printing it (i.e. when an error is encountered).

**Comment by pamaury** [comment]
#18658 was merge and some work is left to a follow-up PR:
- add a mecanism to override the list size on a per-test basis
- add a unittest to make sure that if we push too many items to the list, the code handles it properly.

**Comment by pamaury** [comment]
Yes, I'll create a new PR using some weak linking for the stack so we can call it a day for now.

**Comment by pamaury** [comment]
The last part of this is covered by #19022

**Comment by timothytrippel** [comment]
Looks like this can be closed now.

**Comment by pamaury** [comment]
Reopening the issue. At the moment this is implemented and used by defauly in all OTTF-based tests. One can run the unittest using:
```
./bazelisk.sh test --test_output=streamed --cache_test_results=no //sw/device/tests:status_report_test
```
There are certain aspects that could be considered missing or incomplete:
- the stack trace is only printed if `test_main` returns false but
   - some tests just call `test_status_set` which never returns on error, so this bypasses the stack trace printing
   - `test_main` should really return a `status_t` but because it does not, many tests end up using `CHECK_DIF_OK` that internally uses `test_status_set` on failure, leading to the same problem
- the stack is printed on the UART but ideally should be made available to the host via some ujson commands. Also we have host support for decoding status_t if necessary so together with an ELF file, we can precisely identify the file and line of the error.

Suggestion for future action:
- [ ]  make sure that `test_status_set` prints a stack trace on error
- [ ] when the ottf ujson framework is done, add support for getting the stack trace
- [ ] when ottf ujson framework supports the stack, add stack trace printing and decoding on the host side?

**Comment by pamaury** [comment]
See #20170 for work on this.
